### PR TITLE
Universal CAD artists

### PR DESCRIPTION
**Pull request by tomvanmele**
### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by tomvanmele** [comment]
Based on @gonzalocasas recipe for making nurbs functionality pluggable...

Added `compas.artists.Artist` and pluggable factory `compas.artists.new_artist`, and plugins `compas_rhino.artists.new_artist_rhino` and `compas_blender.artists.new_artist_blender`.

This enables artist instantiation without CAD-specific imports.

```python
from compas.artists import Artist
from compas.datastructures import Mesh
from compas.geometry import Polyhedron

shape = Polyhedron.from_platonicsolid(12)
mesh = Mesh.from_shape(shape)

artist = Artist(mesh)
artist.draw()

print(type(artist))  # compas_rhino.artists.meshartist.MeshArtist (if executed in Rhino)
                     # compas_blender.artists.meshartist.MeshArtist (if executed in Blender)
```

File: src/compas_blender/artists/__init__.py
Comment (gonzalocasas) [inline src/compas_blender/artists/__init__.py:146]: This looks cool as a safe-guard/fail-fast mechanism, but I would consider moving it into the plugin discovery stage, otherwise we will incur this inspection cost on every instantiation of an artist (and also, this is generally a very useful feature to have generalized).   Since moving it to the plugins discovery would need some bigger changes, I think it's ok to just add a `TODO: ` comment here and we go back to it at a later point, especially to not delay this PR much.

Comment (gonzalocasas) [reply src/compas_blender/artists/__init__.py:146]: A quicker alternative would be to move this bit of code into a function like `ensure_abstract_methods_impl`, which does the check but keeps track of classes already checked, so, we can liberally call it anywhere, and it will not be expensive.

File: src/compas_blender/artists/frameartist.py
Comment (gonzalocasas) [inline src/compas_blender/artists/frameartist.py:51]: Just wondering: the `super()` syntax is being intentionally switched to python3 here and everywhere on the `compas_blender` package, right?

File: src/compas_blender/artists/meshartist.py
Comment (gonzalocasas) [inline src/compas_blender/artists/meshartist.py]: not sure here, why is the check for `show_*` not needed anymore?

File: src/compas_blender/artists/meshartist.py
Comment (gonzalocasas) [inline src/compas_blender/artists/meshartist.py]: I don't see where `self.objects` is cleared, does this mean that now calling `.draw_vertices()` twice in a row will end-up drawing duplicated vertices or did I miss the clear somewhere?

File: src/compas_rhino/__init__.py
Comment (gonzalocasas) [inline src/compas_rhino/__init__.py:96]: is this correct? I guess doesn't belong to this technically, but regardless of that.. I believe this was working in the past, right? did rhino install defaults change?

File: src/compas/artists/__init__.py
Comment (gonzalocasas) [inline src/compas/artists/__init__.py]: how about renaming this to `RobotModelArtist` to match all others? (and maybe keep the alias `BaseRobotModelArtist = RobotModelArtist` for backwards compat)

**Comment by beverlylytle** [comment]
Just so you know, I'm currently paused on a PR where I'm adding some Blender artists, waiting for you to finish this. So far I've got point, vector, and line as well as a version of a primitive artist.

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
Generally looking very good, but there's an important missing part here: these also needs to be done for `compas_ghpython` before merging, and that might require some extra logic to figure out how to do the detection on every call to `new_artist`, because a one-time resolution of the plugin will not be enough (user can switch back and forth between GH and Rhino).  Since the plugin system doesn't really have a way to determine which document is active in Rhino (GH's or Rhino's), we might need to have one unified `new_artist` for both that knows this.. or figure out a way to add arbitrary constraints/filters that are tested every single time, not just on discovery.

**Comment by tomvanmele** [comment]
> Just so you know, I'm currently paused on a PR where I'm adding some Blender artists, waiting for you to finish this. So far I've got point, vector, and line as well as a version of a primitive artist.

okay i just added box and sphere.
shall i keep adding shapes and you take care of the primitives?

File: src/compas_blender/artists/frameartist.py
Comment (tomvanmele) [reply src/compas_blender/artists/frameartist.py:51]: yes

**Comment by beverlylytle** [comment]
> shall i keep adding shapes and you take care of the primitives?

sounds good

File: src/compas_rhino/__init__.py
Comment (tomvanmele) [reply src/compas_rhino/__init__.py:96]: we don't really use this anymore, but in any case it is only correct if you have multiple versions of Rhino installed and named one of them `Rhino 7`.

i think we should add an additional option to specify the path of the Rhino installation folder...

File: src/compas_blender/artists/artist.py
Comment (beverlylytle) [inline src/compas_blender/artists/artist.py]: @gonzalocasas  this is where blender objects are cleared.

File: src/compas_blender/artists/meshartist.py
Comment (beverlylytle) [reply src/compas_blender/artists/meshartist.py]: but yes, i think that `clear` will clear out anything that is in `self.objects` and that if `draw_vertices` is called multiple times duplicate objects will be created. but this is behavior is consistent both before and after this pr.

Comment (beverlylytle) [reply src/compas_blender/artists/meshartist.py]: ah, perhaps i misunderstood.  are you suggesting that `self.clear()` is called at the start of any `draw*` method?

**Comment by tomvanmele** [comment]
i have a few pending updates to push that might address the clearing concerns...

File: src/compas/artists/__init__.py
Comment (gonzalocasas) [inline src/compas/artists/__init__.py:47]: Maybe add an internal deprecation note. Not sure if this specific suggestion is the way, but we should come up with a way to tag all these things that we intend to get rid of before 2.0 so that we can do a search-all-files at the right time.

```suggestion
# NOTE: deprecated, will be removed in the next major release
BaseRobotModelArtist = RobotModelArtist
```

File: src/compas/artists/artist.py
Comment (gonzalocasas) [inline src/compas/artists/artist.py:21]: same comment as in #888: maybe these things should not throw `NotImplementedError` but something that clearly signals (both to the user of compas and also to the developer/user reading code to understand) that this is an interface and the plugin is actually implemented (most likely) but just not here. So, something like `compas.plugins.NoPluggableImplFound` or similar.

File: src/compas_blender/artists/__init__.py
Comment (gonzalocasas) [inline src/compas_blender/artists/__init__.py]: Doesn't this registration trigger every time a new artist is created? (Probably side-effect free, but still unnecessary)

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
I added some comments here, but things were not working well on GH, so, I sent a PR to merge into this before we continue reviewing: https://github.com/compas-dev/compas/pull/909

**Comment by gonzalocasas** [comment]
It's hard to really make sure there are no breaking changes in this one because of the size and the fact that all changes happen in non-testable parts of the framework. Maybe after another round of fixes, we could split up and try to manually test groups of features in the team (would also help to list things were we are missing feature parity between the different CADs).

File: src/compas/artists/artist.py
Comment (beverlylytle) [inline src/compas/artists/artist.py:51]: I'm confused why this is on the base `Artist` and not on the `BlenderArtist`.

File: CHANGELOG.md
Comment (beverlylytle) [inline CHANGELOG.md]:
```suggestion
* Added `compas.artist.DataArtistNotRegistered`.
```

File: src/compas/artists/artist.py
Comment (tomvanmele) [reply src/compas/artists/artist.py:51]: ha perhaps the name is confusing. was mean to be a static method to draw many objects at the same time. is not related to the "collection" of blender. but since there is no implementation for it anywhere, perhaps it can just be removed...

File: src/compas_rhino/utilities/drawing.py
Comment (beverlylytle) [inline src/compas_rhino/utilities/drawing.py:91]: I don't like this... if a script is run on mac with the `EditPythonScript` command, then the canvas freezes.  so then, just in `EditPythonScript` one always has to remember to bookend the script with `import compas_rhino .... compas_rhino.redraw()`

Comment (tomvanmele) [reply src/compas_rhino/utilities/drawing.py:91]: i can change it back, but it will not make much difference, because then all of the individual draw functions will set it to false, which is how it was before. just figured it would be less repetitive this way...

Comment (beverlylytle) [reply src/compas_rhino/utilities/drawing.py:91]: mmm... then can there be a `redraw` method on `RhinoArtist`?  it's slightly less annoying to type `artist.redraw()`

Comment (tomvanmele) [reply src/compas_rhino/utilities/drawing.py:91]: i we redraw every draw function, stuff pops up in Rhino 1 object at a time. for example in the case of meshes, this can be really annoying

Comment (tomvanmele) [reply src/compas_rhino/utilities/drawing.py:91]: > mmm... then can there be a `redraw` method on `RhinoArtist`? it's slightly less annoying to type `artist.redraw()`

as a static method?
i could add clear like this as well...

Comment (beverlylytle) [reply src/compas_rhino/utilities/drawing.py:91]: yeah, i agree it's bad to always have redraw enabled

Comment (beverlylytle) [reply src/compas_rhino/utilities/drawing.py:91]: > as a static method?

yes

>i could add clear like this as well...

what would `clear` do?  clear all layers?

Comment (tomvanmele) [reply src/compas_rhino/utilities/drawing.py:91]: clear would be like `compas_rhino.clear()`. it clears everything...

Comment (tomvanmele) [reply src/compas_rhino/utilities/drawing.py:91]:
```python
    @staticmethod
    def clear():
        compas_rhino.clear()

    @staticmethod
    def redraw():
        compas_rhino.redraw()
```

**Review by tomvanmele — COMMENTED** [review]
haha, apparently, because i am the original author of the PR i can't approve the changes, since "pull request authors cannot approve their own pull requests" :)

**Comment by tomvanmele** [comment]
i guess @gonzalocasas this means that you will have to approve of yourself :)

**Review by beverlylytle — APPROVED** [review]
Here I come to save the day!

**Comment by tomvanmele** [comment]
merging still blocked since @gonzalocasas requested changes at some point...

**Review by gonzalocasas — APPROVED** [review]

**Comment by gonzalocasas** [comment]
how picky this thing gets! ;)
I approved the changes, but I could add the missing shape artists to this before we merge.. or shall I do it in a new PR?

**Comment by tomvanmele** [comment]
since @beverlylytle will also still add the primitives for blender in a separate PR i would vote for merging!

**Pull request merged**
